tensorflow/python/keras/utils/io_utils.py

```diff
@@ -33,6 +34,29 @@
   h5py = None
 
 
+if sys.version_info >= (3, 6):
+  import os
+  def path_to_string(path):
+    """
+    If given a non-string typed path object, converts it to its string
+    representation. If the object passed to `path` is not `PathLike` according
+    to https://www.python.org/dev/peps/pep-0519 , then the object is returned
+    unchanged. This allows e.g. passthrough of file objects through this
+    function.
+    """
+    if isinstance(path, os.PathLike):
+      return os.fspath(path)
+    return path
+else:
+  def path_to_string(path):
+    """
+    Dummy implementation of the `path_to_string` function for older versions of
+    python that do not support the file system path protocol
+    https://www.python.org/dev/peps/pep-0519 .
+    """
+    return path
+
+
```

Comment:
I would wrap the `if` with the function instead, hiding the version  switch inside.

```suggestion
def path_to_string(path):
  """Convert a `PathLike` path to a string path, if supported.

  If Python interpreter supports `os.fspath`, then this function works as follows:
  If given a non-string typed path object, converts it to its string
  representation. If the object passed to `path` is not `PathLike` according
  to https://www.python.org/dev/peps/pep-0519 , then the object is returned
  unchanged. This allows e.g. passthrough of file objects through this
  function.
  
  Otherwise,  just return input unchaged, for older versions of
  python that do not support the file system path protocol.

  See https://www.python.org/dev/peps/pep-0519 .
  """
  if sys.version_info >= (3, 6):
    import os
    path = os.fspath(path) if isinstance(path, os.PathLike) else path
    delete os
    return path
  else:
    return path


```

Reply:
I wanted to keep the amount of new code introduced in the code path minimal, and since the python version cannot change after startup, I thought I'd do the test just once.

Your are right though, that in the case of the keras functions covered here, this is highly unlikely to make any performance difference, and anything that interfaces with files is probably many orders of magnitude slower than the version checks.

Why do you think "if in function" is prefereable to "function in if"?

Reply:
I wanted documentation to be in a single place for all codepaths and also to move the `if` execution from import time to the first time the function gets executed. `import tensorflow` is already slow, so we're trying to do as little as possible at import time.

Reply:
I am pretty sure tensorflow or one of its dependencies will load the `os` module at some point, so the import should (as far as my understanding of the import process in python goes) be not much more than a simple dictionary lookup. 

You are right that the documentation looks rather ugly as is.